tests/shared/src/test/scala/matryoshka/example/mathExpr.scala

```diff
@@ -0,0 +1,126 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package matryoshka.example
+
+import org.specs2.mutable._
+
+import slamdata.Predef._
+import matryoshka._
+import matryoshka.data._
+import matryoshka.implicits._
+import scalaz._
+import scala.math.exp
+
+/**
+  * A mathematical expression of a single variable x.
+  *
+  * This can be differentiated using catamorphisms.
+  *
+  * Automatic Differentiation ported from [[https://jtobin.io/ad-via-recursion-schemes Jared Tobin's Automatic Differentiation]]
+  */
+sealed trait MathExprF[A]
+
+object MathExprF {
+  case class VarF[A]() extends MathExprF[A]
+  case class ZeroF[A]() extends MathExprF[A]
+  case class OneF[A]() extends MathExprF[A]
+  case class NegateF[A](a: A) extends MathExprF[A]
+  case class SumF[A](l: A, r: A) extends MathExprF[A]
+  case class ProductF[A](l: A, r: A) extends MathExprF[A]
+  case class ExpF[A](a: A) extends MathExprF[A]
+
+  type Expr = Fix[MathExprF]
+
+  implicit val exprFunctor: Functor[MathExprF] = new Functor[MathExprF] {
+    def map[A, B](fa: MathExprF[A])(f: A => B): MathExprF[B] = fa match {
+      case VarF() => VarF()
+      case ZeroF() => ZeroF()
+      case OneF() => OneF()
+      case NegateF(a) => NegateF(f(a))
+      case SumF(l, r) => SumF(f(l), f(r))
+      case ProductF(l, r) => ProductF(f(l), f(r))
+      case ExpF(a) => ExpF(f(a))
+    }
+  }
+
+  val varExpr: Expr = Fix(VarF())
+  val zero: Expr = Fix(ZeroF())
+  val one: Expr = Fix(OneF())
+  def neg(expr: Expr): Expr = Fix(NegateF(expr))
+  def add(l: Expr, r: Expr): Expr = Fix(SumF( l, r ))
+  def prod(l: Expr, r: Expr): Expr = Fix(ProductF( l, r ))
+  def e(expr: Expr): Expr = Fix(ExpF(expr))
+
+  def evalAlgebra(x: Double): MathExprF[Double] => Double = {
+      case VarF() => x
+      case ZeroF() => 0
+      case OneF() => 1
+      case NegateF(a) => - a
+      case SumF(l, r) => l + r
+      case ProductF(l, r) => l * r
+      case ExpF(a) => exp(a)
+  }
+
+  def eval(x: Double, expr: Expr): Double =
+    expr.cata(evalAlgebra(x))
+
+  def showAlgebra: Algebra[MathExprF, String] = {
+    case VarF() => "x"
+    case ZeroF() => "0"
+    case OneF() => "1"
+    case NegateF(x) => s"-$x"
+    case SumF(l, r) => s"($l + $r)"
+    case ProductF(l, r) => s"($l * $r)"
+    case ExpF(x) => s"e($x)"
+  }
+
+  def diffGAlgebra: GAlgebra[(Expr, ?), MathExprF, Fix[MathExprF]] = {
+    case VarF() => one
+    case ZeroF() => zero
+    case OneF() => zero
+    case NegateF((_, a)) => neg(a)
+    case SumF((_, l), (_, r)) => add(l, r)
+    case ProductF((l, ll), (r, rr)) => add(prod(l, rr),  prod(r, ll))
+    case ExpF((x, xx)) => prod(e(x), xx)
+  }
+
+  def diff(expr: Expr): Expr = expr.para(diffGAlgebra)
+
+  def adAlgebra(x: Double): Algebra[MathExprF, (Double, Double)] = {
```

Comment:
If you change this to a `GAlgebra[(Double, ?), MathExprF, Double]` you could compute the derivative via a zygomorphism in `automaticDifferentiation` using `evalAlgebra` and avoid having to duplicate it here.

Reply:
Thanks!  I was thinking along the same lines, and couldn't quite work out why Jared Tobin's examples didn't do this.